File: tests/Aiko.Core.Tests/ResetCountdownTests.cs
using Aiko.Core;

namespace Aiko.Core.Tests;

public class ResetCountdownTests
{
    private static readonly DateTimeOffset Now = new(2026, 9, 12, 12, 0, 0, TimeSpan.Zero);

    [Fact]
    public void Reports_days_and_hours_when_more_than_a_day_is_left()
    {
        var countdown = ResetCountdown.Between(Now, Now.AddDays(4).AddHours(11).AddMinutes(30));

        Assert.False(countdown.IsReset);
        Assert.Equal(CountdownUnit.DaysAndHours, countdown.Unit);
        Assert.Equal(4, countdown.Days);
        Assert.Equal(11, countdown.Hours);
    }

    [Fact]
    public void Reports_hours_and_minutes_within_a_day()
    {
        var countdown = ResetCountdown.Between(Now, Now.AddHours(3).AddMinutes(34).AddSeconds(50));

        Assert.Equal(CountdownUnit.HoursAndMinutes, countdown.Unit);
        Assert.Equal(0, countdown.Days);
        Assert.Equal(3, countdown.Hours);
        Assert.Equal(34, countdown.Minutes);
    }

    [Fact]
    public void Rounds_down_and_never_promises_more_time()
    {
        var countdown = ResetCountdown.Between(Now, Now.AddMinutes(12).AddSeconds(59));

        Assert.Equal(CountdownUnit.Minutes, countdown.Unit);
        Assert.Equal(12, countdown.Minutes);
    }

    [Fact]
    public void Reports_less_than_a_minute_when_almost_no_time_is_left()
    {
        var countdown = ResetCountdown.Between(Now, Now.AddSeconds(59));

        Assert.False(countdown.IsReset);
        Assert.Equal(CountdownUnit.LessThanMinute, countdown.Unit);
    }

    [Fact]
    public void Exactly_one_day_is_still_days_and_hours()
    {
        var countdown = ResetCountdown.Between(Now, Now.AddDays(1));

        Assert.Equal(CountdownUnit.DaysAndHours, countdown.Unit);
        Assert.Equal(1, countdown.Days);
        Assert.Equal(0, countdown.Hours);
    }

    [Fact]
    public void A_window_whose_time_has_passed_is_reset()
    {
        var countdown = ResetCountdown.Between(Now, Now.AddMinutes(-1));

        Assert.True(countdown.IsReset);
    }

    [Fact]
    public void The_moment_of_reset_counts_as_reset()
    {
        var countdown = ResetCountdown.Between(Now, Now);

        Assert.True(countdown.IsReset);
    }
}

File: mac/Tests/AikoKitTests/ResetCountdownTests.swift
import Foundation
import Testing

@testable import AikoKit

struct ResetCountdownTests {
    static let now = utc(2026, 9, 12, 12)

    @Test
    func reportsDaysAndHoursWhenMoreThanADayIsLeft() {
        let countdown = ResetCountdown.between(
            Self.now, Self.now.adding(days: 4).adding(hours: 11).adding(minutes: 30))

        #expect(!countdown.isReset)
        #expect(countdown.unit == .daysAndHours)
        #expect(countdown.days == 4)
        #expect(countdown.hours == 11)
    }

    @Test
    func reportsHoursAndMinutesWithinADay() {
        let countdown = ResetCountdown.between(
            Self.now, Self.now.adding(hours: 3).adding(minutes: 34).adding(seconds: 50))

        #expect(countdown.unit == .hoursAndMinutes)
        #expect(countdown.days == 0)
        #expect(countdown.hours == 3)
        #expect(countdown.minutes == 34)
    }

    @Test
    func roundsDownAndNeverPromisesMoreTime() {
        let countdown = ResetCountdown.between(Self.now, Self.now.adding(minutes: 12).adding(seconds: 59))

        #expect(countdown.unit == .minutes)
        #expect(countdown.minutes == 12)
    }

    @Test
    func reportsLessThanAMinuteWhenAlmostNoTimeIsLeft() {
        let countdown = ResetCountdown.between(Self.now, Self.now.adding(seconds: 59))

        #expect(!countdown.isReset)
        #expect(countdown.unit == .lessThanMinute)
    }

    @Test
    func exactlyOneDayIsStillDaysAndHours() {
        let countdown = ResetCountdown.between(Self.now, Self.now.adding(days: 1))

        #expect(countdown.unit == .daysAndHours)
        #expect(countdown.days == 1)
        #expect(countdown.hours == 0)
    }

    @Test
    func aWindowWhoseTimeHasPassedIsReset() {
        let countdown = ResetCountdown.between(Self.now, Self.now.adding(minutes: -1))

        #expect(countdown.isReset)
    }

    @Test
    func theMomentOfResetCountsAsReset() {
        let countdown = ResetCountdown.between(Self.now, Self.now)

        #expect(countdown.isReset)
    }
}
